upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+      UPB_UNREACHABLE();
+    }
+    *msg_modifiers = mod;
+  } else {
+    upb_MiniTable_Field* field = &d->fields[field_count - 1];
+    upb_MtDecoder_ModifyField(d, mod, field);
+  }
+
+  return ptr;
+}
+
+static void upb_MtDecoder_AllocateSubs(upb_MtDecoder* d, uint32_t sub_count) {
+  size_t subs_bytes = sizeof(*d->table->subs) * sub_count;
+  d->table->subs = upb_Arena_Malloc(d->arena, subs_bytes);
+  upb_MtDecoder_CheckOutOfMemory(d, d->table->subs);
+}
+
+static void upb_MtDecoder_Parse(upb_MtDecoder* d, const char* ptr, size_t len,
+                                void* fields, size_t field_size,
+                                uint16_t* field_count, uint32_t* sub_count) {
+  uint64_t msg_modifiers = 0;
+  uint32_t last_field_number = 0;
+  bool need_dense_below = d->table != NULL;
+
+  d->end = ptr + len;
+
+  while (ptr < d->end) {
+    char ch = *ptr++;
+    if (ch <= kUpb_EncodedValue_MaxField) {
+      upb_MiniTable_Field* field = fields;
+      *field_count += 1;
+      fields = (char*)fields + field_size;
+      field->number = ++last_field_number;
+      upb_MiniTable_SetField(d, ch, field, msg_modifiers, sub_count);
+    } else if (kUpb_EncodedValue_MinModifier <= ch &&
+               ch <= kUpb_EncodedValue_MaxModifier) {
+      ptr = upb_MtDecoder_ParseModifier(d, ptr, ch, *field_count, &msg_modifiers);
+    } else if (ch == kUpb_EncodedValue_End) {
+      if (!d->table) {
+        upb_MtDecoder_ErrorFormat(d, "Extensions cannot have oneofs.");
+        UPB_UNREACHABLE();
+      }
+      ptr = upb_MtDecoder_DecodeOneofs(d, ptr);
+    } else if (kUpb_EncodedValue_MinSkip <= ch &&
+               ch <= kUpb_EncodedValue_MaxSkip) {
+      if (need_dense_below) {
+        d->table->dense_below = d->table->field_count;
+        need_dense_below = false;
+      }
+      uint32_t skip;
+      ptr = upb_MiniTable_DecodeBase92Varint(d, ptr, ch,
+                                             kUpb_EncodedValue_MinSkip,
+                                             kUpb_EncodedValue_MaxSkip, &skip);
+      last_field_number += skip;
+      last_field_number--;  // Next field seen will increment.
+    }
+  }
+
+  if (need_dense_below) {
+    d->table->dense_below = d->table->field_count;
+  }
+}
+
+static void upb_MtDecoder_ParseMessage(upb_MtDecoder* d, const char* data,
+                                       size_t len) {
+  // Buffer length is an upper bound on the number of fields. We will return
+  // what we don't use.
+  d->fields = upb_Arena_Malloc(d->arena, sizeof(*d->fields) * len);
+  upb_MtDecoder_CheckOutOfMemory(d, d->fields);
+
+  uint32_t sub_count = 0;
+  d->table->field_count = 0;
+  d->table->fields = d->fields;
+  upb_MtDecoder_Parse(d, data, len, d->fields, sizeof(*d->fields),
+                      &d->table->field_count, &sub_count);
+
+  // Return unused memory from fields array.
+  assert(d->table->field_count <= len);
+  d->fields = upb_Arena_Realloc(d->arena, d->fields, sizeof(*d->fields) * len,
```

Comment:
I have a mild worry that this will actually cause a double allocation because something deep in the guts of Parse used the arena to allocate again.  Can you put in an assert or something to make sure this optimization doesn't age into a pessimization?

Reply:
Done.